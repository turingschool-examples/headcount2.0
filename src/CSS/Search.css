.search-bar {
  margin: 1rem 0.5rem;
}

.search-input {
  height: 2rem;
  width: 20rem;
  border: 1.5px solid #ccc;
  margin: 1rem 0;
  border-radius: 5px;
}

.search-input:hover {
  box-shadow: 1px 1px 1px 1px #ccc;
}

.hide-suggestions {
  display: none;
}

.did-you-mean {
  font-style: italic;
  color: gray;
}

.show-suggestions {
  margin: 1rem 0;
  text-align: center;
}

.suggestions {
  cursor: pointer;
  border: 1px solid gray;
  margin: 0;
  padding: 2px;
  text-align: center;
}

.suggestions:hover {
  background-color: rgb(244, 243, 243);
  box-shadow: 1.5px 1.5px 1.5px 1.5px #ccc;
}

button {
  width: 21.3rem;
  height: 2rem;
  border-radius: 5px;
  background-color: rgb(230, 230, 230);
  font-size: 1.1rem;
}

button:hover {
  background-color: rgb(142, 142, 142);
  color: rgb(237, 237, 237);
  cursor: pointer;
}

input[type="text"] {
  font-size: 1.1rem;
  line-height: 5rem;
  padding-left: 1rem;
}

.show {
  margin-bottom: 10rem;
  font-style: italic;
}

.show-none {
  display: none;
}

.display-error {
  padding: 1rem 0;
  text-align: center;
}
