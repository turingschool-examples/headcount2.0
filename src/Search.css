.search {
  text-align: center
}

.inputField {
  padding: 10px;
  border: 3px solid red;
  /*margin: 10px;*/
  width: 300px;
}

.inputButton {
  padding: 10px;
  background-color: white;
  margin: 10px;
  border-radius: 30%;
  border: 3px solid green;
  color: green;
}

.inputButton:hover {
  background-color: green;
  color: white;
}
