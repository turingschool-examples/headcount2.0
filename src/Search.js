import React, { Component } from "react";
import DistrictRepository from "./helper";
import kindergarners from "./data/kindergartners_in_full_day_program";

export default class Search extends Component {
  constructor(props) {
    super(props);
    this.state = {
      userInput: "",
      searchSuggestions: []
    };
    this.districtRepository = new DistrictRepository(kindergarners);
  }

  handleChange = ({ name, value }) => {
    this.setState({ [name]: value });
    this.suggestDistricts(this.state.userInput);
  };

  suggestDistricts = e => {
    let suggestions = this.districtRepository.findAllMatches(
      this.state.userInput
    );
    this.setState({ searchSuggestions: suggestions });
    if (e.length > 2) {
      this.props.handleSubmit(this.state.searchSuggestions);
    }
  };

  searchDistrict = district => {
    const filteredDistricts = this.districtRepository.findByName(district);
    this.props.handleSubmit([filteredDistricts]);
  };

  render() {
    return (
      <div>
        <form>
          <input
            placeholder="Search District"
            name="userInput"
            onChange={e => this.handleChange(e.target)}
            value={this.state.userInput}
          />
          {/* <button onClick={e => this.searchDistrict(e)}>Search</button> */}
        </form>
        <div>
          {this.state.searchSuggestions.map((district, i) => {
            if (
              this.state.searchSuggestions &&
              this.state.userInput.length > 3
            ) {
              return (
                <p
                  className="suggestions"
                  key={i}
                  onClick={e => this.searchDistrict(e.target.textContent)}
                >
                  {district.location}
                </p>
              );
            }
          })}
        </div>
      </div>
    );
  }
}
