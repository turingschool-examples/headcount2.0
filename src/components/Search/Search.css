/*SEARCH COMPONENT RELATED CSS GOES HERE*/
.Search {
  width: 35%;
  display: flex;
  justify-content: space-between
}

input {
  height: 2rem;
  width: 100%;
  border: 2px solid #BEBEBE;
  padding-left: .6rem;
  font-size: 1.25rem;
  letter-spacing: 0.5px;
  color: #006989;
}

::placeholder {
  font-family: "Roboto", sans-serif;
  font-weight: 400;
  color: #9E9E9E;
}