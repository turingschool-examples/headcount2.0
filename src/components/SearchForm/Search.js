import React, { Component } from 'react';
import PropTypes from 'prop-types';

export class Search extends Component {
  constructor() {
    super();
    this.state = {
      query: ''
    };
  }

  handleChange =  event => {
    const { value } = event.target;

    this.props.filterDistricts(value);

    this.setState({ query: value });
  }

  render() {
    const { query } = this.state;

    return (
      <input
        onChange={this.handleChange}
        type="text"
        name="query"
        value={query}
        placeholder="Search"
      />
    );
  }
}

Search.propTypes = {
  filterDistricts: PropTypes.func.isRequired
};

export default Search;
