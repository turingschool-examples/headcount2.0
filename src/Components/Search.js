import React, { Component } from "react";
import DistrictRepository from "../helper";
import kindergarners from "../data/kindergartners_in_full_day_program";
import PropTypes from "prop-types";
import "../CSS/Search.css";

export default class Search extends Component {
  constructor(props) {
    super(props);
    this.state = {
      districtInputOne: "",
      searchSuggestions: []
    };
    this.districtRepository = new DistrictRepository(kindergarners);
  }

  handleChange = ({ value }) => {
    this.setState({ districtInputOne: value });
    this.suggestDistricts();
  };

  suggestDistricts = () => {
    const { districtInputOne } = this.state;
    const searchSuggestions = this.districtRepository.findAllMatches(
      districtInputOne
    );
    this.props.selectedDistricts.forEach(district => {
      searchSuggestions.push(district);
    });

    this.setState({ searchSuggestions });
    if (districtInputOne.length > 1) {
      this.props.handleSubmit(searchSuggestions);
    } else {
      const newDistricts = this.districtRepository.districtsArray();
      this.props.handleSubmit(newDistricts);
    }
  };

  clearInput = () => {
    this.setState({ districtInputOne: "" });
    const newDistricts = this.districtRepository.districtsArray();
    this.props.handleSubmit(newDistricts);
  };

  searchDistrict = district => {
    const filteredDistricts = this.districtRepository.findByName(district);
    this.props.handleSubmit([filteredDistricts]);
  };

  render() {
    return (
      <div className="search-bar">
        <form>
          <input
            type="text"
            className="search-input"
            placeholder={
              this.props.selectedDistricts.length > 0
                ? "Remove Card to Search"
                : "Search District"
            }
            onChange={event => this.handleChange(event.target)}
            value={this.state.districtInputOne}
            disabled={this.props.selectedDistricts.length > 1}
          />
          <button onClick={() => this.props.clearComparisons()}>
            Clear Fields
          </button>
        </form>
        <div
          className={
            this.state.districtInputOne.length > 1
              ? "show-suggestions"
              : "hide-suggestions"
          }
        >
          <span
            className={
              this.state.searchSuggestions.length
                ? "show-none"
                : "display-error"
            }
          >
            None Found! Search Again
          </span>

          <span
            className={
              this.state.searchSuggestions.length ? "show" : "show-none"
            }
          >
            Did you Mean?
          </span>
          {this.state.searchSuggestions.map((district, index) => {
            if (
              this.state.searchSuggestions &&
              this.state.districtInputOne.length > 1
            ) {
              var suggestions = (
                <p
                  className="suggestions"
                  key={index}
                  onClick={event => {
                    this.props.selectCard(event.target.textContent);
                    this.clearInput();
                  }}
                >
                  {district.location}
                </p>
              );
            }
            return suggestions;
          })}
        </div>
      </div>
    );
  }
}

const { func, arrayOf, shape, bool, string, number } = PropTypes;

Search.propTypes = {
  selectedDistricts: arrayOf(
    shape({
      clicked: bool,
      location: string,
      stats: shape({ year: number })
    })
  ),
  handleSubmit: func.isRequired,
  selectCard: func.isRequired,
  clearComparisons: func.isRequired
};
