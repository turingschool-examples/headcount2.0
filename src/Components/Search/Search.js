import React from 'react';
import PropTypes from 'prop-types';
import './Search.css';

const Search = ( {handleSearch} ) => {
  return (
      <div className="search-container">
        <input 
        placeholder="Search for district..."
        className="search"
        onChange={ (e) => handleSearch(e.target.value) }/>
      </div>
    )
}

Search.propTypes = {
  handleSearch: PropTypes.func.isRequired
}

export default Search;