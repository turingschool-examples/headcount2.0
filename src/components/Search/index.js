import React, { Component } from 'react'
import PropTypes from 'prop-types'
import './Search.css'

class Search extends Component {
  constructor({findAllMatches}) {
    super()
    this.state = {
      query: ''
    }
  }

  handleInputChange = event => {
    const { value } = event.target

    this.setState({query: value})
    this.props.findAllMatches(value);
  }

  handleSubmit = event => {
    event.preventDefault()
  }

  render() {
    return(
      <form className="Search" onSubmit={this.handleSubmit} >
        <input
          className="search-input"
          aria-label="Search for a specific district here"
          placeholder="Search for a specific district"
          name="query"
          value={this.state.query}
          onChange={this.handleInputChange}
        />
      </form>
    )
  }
}

Search.propTypes = {
  findAllMatches: PropTypes.func.isRequired
}

export default Search