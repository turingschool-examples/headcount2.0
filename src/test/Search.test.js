import React from "react";
import ReactDOM from "react-dom";
import Search from "../Components/Search";

describe("Search", () => {
  let shallowWrap;
  let mockSubmit = jest.fn();
  let mockSelect = jest.fn();
  let mockClear = jest.fn();
  let mockData = [
    {
      clicked: true,
      location: "ADAMS COUNTY 14",
      stats: { 2004: 0.24, 2005: 0.278, 2006: 0.337, 2007: 0.395 }
    },
    {
      clicked: true,
      location: "COLORADO",
      stats: { 2004: 0.24, 2005: 0.278, 2006: 0.337, 2007: 0.395 }
    }
  ];

  beforeEach(() => {
    shallowWrap = shallow(
      <Search
        selectedDistricts={mockData}
        handleSubmit={() => mockSubmit()}
        selectCard={e => mockSelect(e)}
        clearComparisons={() => mockClear()}
      />
    );
  });

  it("should exist", () => {
    expect(shallowWrap).toBeDefined;
  });

  it("should have default state properties", () => {
    expect(shallowWrap.state().districtInputOne).toEqual("");
    expect(shallowWrap.state().searchSuggestions).toEqual([]);
  });

  it("should handleChange; set State and call suggest method", () => {
    shallowWrap.instance().handleChange({ value: "COLORADO" });

    expect(shallowWrap.state().districtInputOne).toEqual("COLORADO");
  });

  it("should set state of Search suggestions based on userInput", () => {
    shallowWrap.instance().handleChange({ value: "Colorado" });

    const result = [
      {
        clicked: false,
        location: "COLORADO",
        stats: {
          "2004": 0.24,
          "2005": 0.278,
          "2006": 0.337,
          "2007": 0.395,
          "2008": 0.536,
          "2009": 0.598,
          "2010": 0.64,
          "2011": 0.672,
          "2012": 0.695,
          "2013": 0.703,
          "2014": 0.741
        }
      },
      {
        clicked: false,
        location: "COLORADO SPRINGS 11",
        stats: {
          "2004": 0.069,
          "2005": 0.509,
          "2006": 0.638,
          "2007": 0.994,
          "2008": 0.992,
          "2009": 1,
          "2010": 0.993,
          "2011": 0.994,
          "2012": 0.993,
          "2013": 0.989,
          "2014": 0.994
        }
      }
    ];

    shallowWrap.instance().suggestDistricts("Colorado");
    expect(shallowWrap.state().searchSuggestions).toEqual(result);
  });

  it("should send call handleSubmit in App", () => {
    shallowWrap.instance().searchDistrict("COLORADO");
    expect(mockSubmit).toHaveBeenCalled();
  });

  it("should render form container with input", () => {
    expect(shallowWrap.find("form").length).toEqual(1);
    expect(shallowWrap.find("input").length).toEqual(1);
  });

  it("should change state onChange of input field", () => {
    let eventOne = { target: { value: "Academy" } };
    let eventTwo = { target: { value: "Akron" } };

    let userInput = shallowWrap.find("input");
    userInput.simulate("change", eventOne);
    expect(shallowWrap.state().districtInputOne).toEqual("Academy");

    userInput.simulate("change", eventTwo);
    expect(shallowWrap.state().districtInputOne).toEqual("Akron");
  });

  it("should suggest districts based on user input", () => {
    let eventOne = { target: { value: "Acade" } };
    let event = { target: { textContent: "ACADEMY 20" } };
    let mockSelect = jest.fn();
    let mockSubmit = jest.fn();
    let mockClear = jest.fn();

    let shallowWrap = shallow(
      <Search
        selectedDistricts={mockData}
        handleSubmit={mockSubmit}
        selectCard={mockSelect}
        clearComparisons={mockClear}
      />
    );

    let mountWrap = mount(
      <Search
        selectedDistricts={mockData}
        handleSubmit={mockSubmit}
        selectCard={mockSelect}
        clearComparisons={mockClear}
      />
    );

    let userInput = shallowWrap.find("input");

    userInput.simulate("change", eventOne);
    let suggestionOne = shallowWrap.find("p");

    let suggestedOneContent = suggestionOne.props().children;
    expect(suggestedOneContent).toEqual("ACADEMY 20");

    let userInputTwo = mountWrap.find("input");
    userInputTwo.simulate("change", eventOne);

    let suggestionTwo = mountWrap.find("p").at(0);
    suggestionTwo.simulate("click");
    expect(mockSelect).toHaveBeenCalled();
  });

  it("should clear input, invoke the handleSubmit method from state", () => {
    let mockSelect = jest.fn();
    let mockSubmit = jest.fn();
    let mockClear = jest.fn();

    let shallowWrap = shallow(
      <Search
        selectedDistricts={mockData}
        handleSubmit={mockSubmit}
        selectCard={mockSelect}
        clearComparisons={mockClear}
      />
    );

    shallowWrap.setState({ districtInputOne: "Denver" });
    shallowWrap.instance().clearInput();
    expect(shallowWrap.state().districtInputOne).toEqual("");
  });

  it("should search for specific district and invoke the handleSubmit Method from state", () => {
    shallowWrap.instance().searchDistrict("ACADEMY 20");
    expect(mockSubmit).toHaveBeenCalled();
  });

  it("should match snapshot", () => {
    expect(shallowWrap).toMatchSnapshot();
  });

  it("should match snapshot", () => {
    shallowWrap.instance().searchDistrict;
    expect(shallowWrap).toMatchSnapshot();
  });
});
