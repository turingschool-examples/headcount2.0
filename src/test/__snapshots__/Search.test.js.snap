// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should match snapshot 1`] = `
<div
  className="search-bar"
>
  <form>
    <input
      className="search-input"
      disabled={true}
      onChange={[Function]}
      placeholder="Remove Card to Search"
      type="text"
      value=""
    />
    <button
      onClick={[Function]}
    >
      Clear Fields
    </button>
  </form>
  <div
    className="hide-suggestions"
  >
    <span
      className="display-error"
    >
      None Found! Search Again
    </span>
    <span
      className="show-none"
    >
      Did you Mean?
    </span>
  </div>
</div>
`;

exports[`Search should match snapshot 2`] = `
<div
  className="search-bar"
>
  <form>
    <input
      className="search-input"
      disabled={true}
      onChange={[Function]}
      placeholder="Remove Card to Search"
      type="text"
      value=""
    />
    <button
      onClick={[Function]}
    >
      Clear Fields
    </button>
  </form>
  <div
    className="hide-suggestions"
  >
    <span
      className="display-error"
    >
      None Found! Search Again
    </span>
    <span
      className="show-none"
    >
      Did you Mean?
    </span>
  </div>
</div>
`;
